quaddtype/quaddtype/src/casts.cpp

```diff
@@ -0,0 +1,449 @@
+#include "dtype.h"
+
+static NPY_CASTING
+quad_to_quad_resolve_descriptors(PyObject *NPY_UNUSED(self),
+                                 PyArray_DTypeMeta *NPY_UNUSED(dtypes[2]),
+                                 QuadPrecDTypeObject *given_descrs[2],
+                                 QuadPrecDTypeObject *loop_descrs[2], npy_intp *view_offset)
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+
+    if (given_descrs[1] == NULL) {
+        Py_INCREF(given_descrs[0]);
+        loop_descrs[1] = given_descrs[0];
+    }
+    else {
+        Py_INCREF(given_descrs[1]);
+        loop_descrs[1] = given_descrs[1];
+    }
+
+    *view_offset = 0;
+    return NPY_SAME_KIND_CASTING;
+}
+
+static int
+quad_to_quad_strided_loop(PyArrayMethod_Context *context, char *const data[],
+                          npy_intp const dimensions[], npy_intp const strides[],
+                          void *NPY_UNUSED(auxdata))
+{
+    npy_intp N = dimensions[0];
+    char *in_ptr = data[0];
+    char *out_ptr = data[1];
+
+    while (N--) {
+        Sleef_quad *in = (Sleef_quad *)in_ptr;
+        Sleef_quad *out = (Sleef_quad *)out_ptr;
+
+        *out = *in;
+
+        in_ptr += strides[0];
+        out_ptr += strides[1];
+    }
+    return 0;
+}
+
+// Casting from other types to QuadDType
+
+template <typename T>
+static inline Sleef_quad
+to_quad(T x);
+
+template <>
+inline Sleef_quad
+to_quad<npy_bool>(npy_bool x)
+{
+    return x ? Sleef_cast_from_doubleq1(1.0) : Sleef_cast_from_doubleq1(0.0);
+}
+template <>
+inline Sleef_quad
+to_quad<npy_byte>(npy_byte x)
+{
+    return Sleef_cast_from_int64q1(x);
+}
+// template <>
+// inline Sleef_quad
+// to_quad<npy_ubyte>(npy_ubyte x)
+// {
+//     return Sleef_cast_from_uint64q1(x);
+// }
+template <>
+inline Sleef_quad
+to_quad<npy_short>(npy_short x)
+{
+    return Sleef_cast_from_int64q1(x);
+}
+template <>
+inline Sleef_quad
+to_quad<npy_ushort>(npy_ushort x)
+{
+    return Sleef_cast_from_uint64q1(x);
+}
+template <>
+inline Sleef_quad
+to_quad<npy_int>(npy_int x)
+{
+    return Sleef_cast_from_int64q1(x);
+}
+template <>
+inline Sleef_quad
+to_quad<npy_uint>(npy_uint x)
+{
+    return Sleef_cast_from_uint64q1(x);
+}
+template <>
+inline Sleef_quad
+to_quad<npy_long>(npy_long x)
+{
+    return Sleef_cast_from_int64q1(x);
+}
+template <>
+inline Sleef_quad
+to_quad<npy_ulong>(npy_ulong x)
+{
+    return Sleef_cast_from_uint64q1(x);
+}
+template <>
+inline Sleef_quad
+to_quad<npy_longlong>(npy_longlong x)
+{
+    return Sleef_cast_from_int64q1(x);
+}
+template <>
+inline Sleef_quad
+to_quad<npy_ulonglong>(npy_ulonglong x)
+{
+    return Sleef_cast_from_uint64q1(x);
+}
+template <>
+inline Sleef_quad
+to_quad<float>(float x)
+{
+    return Sleef_cast_from_doubleq1(x);
+}
+template <>
+inline Sleef_quad
+to_quad<double>(double x)
+{
+    return Sleef_cast_from_doubleq1(x);
+}
+template <>
+inline Sleef_quad
+to_quad<long double>(long double x)
+{
+    return Sleef_cast_from_doubleq1(x);
```

Comment:
I am not sure here, but as handling earlier we can use `memcpy` to copy the `long double` value to an aligned buffer before passing it to the `Sleef_cast_from_doubleq1`
that way it can handle the non-alignment, something like this
```c++
alignas(long double) long double aligned_x;
memcpy(&aligned_x, &x, sizeof(long double));
return Sleef_cast_from_doubleq1(aligned_x);
```